### PR TITLE
Correctly replace scene root when `must_reload` in `EditorData::check_and_update_scene()`

### LINKED ISSUE #90301: Saving sub-scenes will mess up the display of the main scene on the 2D screen
### Tested versions

v4.3.dev5.official [c9c17d6ca] good
f47f4a02c bad

### System information

Linux Mint 21.3 (Virginia) - X11 - Vulkan (Forward+) - dedicated NVIDIA GeForce GTX 1050 Ti (nvidia; 535.161.07) - Intel(R) Core(TM) i5-7300HQ CPU @ 2.50GHz (4 Threads)

### Issue description

Saving the sub-scene and switching to the main scene reloads the scene, and it seems like the position calculation of something is dependent on the current camera.

If there is a **enabled** (default value) `Camera2D` node in the main scene, some things in the main scene will be displayed in the wrong position after reloading. For example, the position of the camera screen rectangle. `Parallax2D`'s `screen_offset` may also change.

https://github.com/godotengine/godot/assets/30386067/869cba0c-e815-47b0-a7dc-5fad8c0f11fa

Closing and reopening the scene will display correctly.

### Steps to reproduce

1. Download and use godot to open the MRP;
2. Open main.tscn;
3. Open sub.tscn, and save;
4. Switch to the tab of main.tscn, it will be automatically reloaded, and the displayed content will be messed up after reloading.

### Minimal reproduction project (MRP)

[wrong_display.zip](https://github.com/godotengine/godot/files/14893243/wrong_display.zip)

### PR DESCRIPTION
**Pull request by Rindbee**
We need to update the scene root in multiple singletons to ensure that certain flags are correct. This is what `EditorNode::set_edited_scene()` does.
    
Usually we use `replace_by` to complete the replacement of the scene root. Call `EditorNode::set_edited_scene()` when the `replacing_by` signal is emitted to set the new scene root. This is suitable when using a single node to replace, but this may be problematic if the replacing node is a tree. Because during the call to `replace_by()`, the new node and its child nodes will enter tree during `parent->add_child(p_node)`, and later emits the `replacing_by` signal. This results in some flags potentially being incorrect during the enter tree of the new node's children.
    
When the parent scene has to be reloaded because the child scene changes and switches to the parent scene, there is no need to use `replace_by()` since the scene's diffs are already saved.

Fix #90301.

### CODE REVIEW

**Comment by akien-mga** [comment]
CC @ajreckof

I feel like we're going around in circles there, we modified `replace_by` for this but now it's no longer needed.
Might be best to go back to the original state and figure out again how to solve the issue without regression.

**Comment by AThousandShips** [comment]
It does feel like where playing a game of whack-a-bug with this issue, so it'd indeed be good to revert to a clean slate and dig deeper figure out what's causing what, and make sure we have a fix that doesn't cause regression, it seems like some part of the process of replacement here is missed and that it is causing these problems, due to things occuring behind the scenes that were missed

**Comment by Rindbee** [comment]
Maybe the timing I chose for `replacing_by` in #68599 was inappropriate. Emitting it a little early helps ensure that certain types of nodes are configured correctly when entering the tree as the scene root.

File: scene/main/node.cpp [hunk 1]
```diff
@@ -3027,51 +3026,48 @@ void Node::replace_by(Node *p_node, bool p_keep_groups, bool p_keep_children) {
...
+		for (int i = 0; i < get_child_count(); i++) {
+			find_owned_by(data.owner, get_child(i), &owned_by_owner);
 		}
+
 		_clean_up_owner();
 	}
-	Node *parent = data.parent;
-	int index_in_parent = get_index(false);
 
+	emit_signal(SNAME("replacing_by"), p_node);
```
[the remark this answers is not in the document]
Comment (Rindbee) [reply scene/main/node.cpp:3036]: Still need in situations where the scene root changes type. 

Before the node enters the tree, the cache of the scene root node needs to be updated (call `EditorNode::set_edited_scene()`), that is, the new scene root needs to be determined before `parent->add_child(p_node);` (line 3042).

In the editor, `replacing_by` is only used in `EditorNode::set_edited_scene()`.

**Review by KoBeWi — APPROVED** [review]
Looks fine.
The third commit should likely be merged with the first one.

File: doc/classes/Node.xml [hunk 2]
```diff
@@ -1062,7 +1061,7 @@
 			<param index="0" name="node" type="Node" />
 			<description>
 				Emitted when this node is being replaced by the [param node], see [method replace_by].
-				This signal is emitted [i]after[/i] [param node] has been added as a child of the original parent node, but [i]before[/i] all original child nodes have been reparented to [param node].
+				This signal is emitted [i]before[/i] [param node] will be added as a child of the original parent node.
```
Comment (AThousandShips) [inline doc/classes/Node.xml:1064]:
```suggestion
				This signal is emitted [i]before[/i] [param node] is be added as a child of the original parent node.
```

**Comment by ajreckof** [comment]
The third commit won't cause problems in the editor but since it is a signal hence exposed this is still a compatibility breakage. I'm not sure the bug this specifically fix warrant this compatibility breakage. Whatever I think it is better to keep separate if we finally decide to revert it because it was in the end too much of a compatibility breakage.
For us to have a better appreciation of what it fixes and if it is worth the compatibility breakage, it would be great to have a clear reproduction step to a bug that would be fixed by this.
I'm wondering if it shouldn't have it's own PR for those reasons.

File: doc/classes/Node.xml (hunk 2 above)
Comment (KoBeWi) [reply doc/classes/Node.xml:1064]:
```suggestion
				This signal is emitted [i]before[/i] [param node] is added as a child of the original parent node.
```

**Comment by Rindbee** [comment]
Create PR #90479 for the third commit.

**Review by akien-mga — APPROVED** [review]

**Review by akien-mga — APPROVED** [review]

**Comment by akien-mga** [comment]
Thanks!

**Pull request merged**